File: Swift/BankaiCore/Sources/BankaiCore/ARepositoryOf.swift
//
//  ARepositoryOf.swift
//
//
//  Created by Sergio Daniel on 1/09/24.
//

// Complements with: RepositoryCore
public struct ARepositoryOf<V, F, L, R>: Repository
    where V: Traceable, F: Filter, L: LocalDataSource, R: RemoteDataSource,
        L.FilterType == F, L.ValueType == V, R.FilterType == F, R.ValueType == V
{
    public typealias FilterType = F
    public typealias ValueType = V
    public typealias Local = L
    public typealias Remote = R
    
    public var local: L
    public var remotes: [R]
    
    public init(local: L) {
        self.local = local
        self.remotes = []
    }
    
    public mutating func prepare() {
        local.prepare()
        
        for i in 0..<remotes.count {
            remotes[i].prepare()
        }
    }
    
    public mutating func add(remote: R) {
        remotes.append(remote)
    }
}

File: C#/BankaiCore/BankaiCore/Repository/ARepositoryOf.cs
using System;
using System.Collections.Generic;
using System.Formats.Asn1;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BankaiCore.Repository;

public class ARepositoryOf<T, F>(LocalDataSource<T, F> local) : Repository<T, F>
    where T : Traceable
    where F : Filter
{
    public LocalDataSource<T, F> local { get; init; } = local;
    public List<RemoteDataSource<T, F>> remotes { get; set; } = new();

    public async Task prepare()
    {
        await local.prepare();
        foreach (var remote in remotes)
        {
            await remote.prepare();
        }
    }

    public async Task<List<T>> Read(F? filter)
    {
        List<T> items = await local.retrieve(filter);
        return items;
    }

    public async Task Commit(T item)
    {
        await local.store([ item ]);
    }

    public async Task<List<T>> pull(F? filter)
    {
        List<T> items = await local.retrieve(filter);
        foreach (var remote in remotes)
        {
            items.AddRange(await remote.pull(filter));
        }
        return items;
    }

    public async Task push(List<T> items)
    {
        await local.store(items);
        foreach (var remote in remotes)
        {
            await remote.push(items);
        }
    }
}